### PR TITLE
Check Java Toolchain install folder for left-overs

### PR DESCRIPTION
**Pull request by jbartok**
Relates to #22803

### CODE REVIEW

**Review by ljacomet — CHANGES_REQUESTED** [review]
Check makes sense and detecting the metadata in such a corner case is a good idea.

Request about the exception message though, see below

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java [hunk 1]
```diff
@@ -160,6 +161,32 @@ public File provisionFromArchive(JavaToolchainSpec spec, File jdkArchive, URI ur
...
+        }
+
+        String leftoverMetadata;
+        try {
+            leftoverMetadata = getMetadata(markerLocation).toString();
+        } catch (Exception e) {
+            leftoverMetadata = "Could not be determined due to: " + e.getMessage();
+        }
+        LOGGER.warn("While provisioning Java toolchain from '{}' to satisfy spec '{}', having metadata '{}', " +
+                "leftover content with metadata '{}' found in install folder '{}'. Will be erased, but this should not have happened.",
```
[the remark this answers is not in the document]
Comment (jbartok) [reply subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java]: Done

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java [hunk 2]
```diff
@@ -160,6 +161,32 @@ public File provisionFromArchive(JavaToolchainSpec spec, File jdkArchive, URI ur
...
+        File markerLocation = markedLocation(installFolder);
+        if (!isMarkedLocation(markerLocation)) {
+            return; //no marker found
+        }
+
+        String leftoverMetadata;
+        try {
+            leftoverMetadata = getMetadata(markerLocation).toString();
+        } catch (Exception e) {
+            leftoverMetadata = "Could not be determined due to: " + e.getMessage();
```
Comment (octylFractal) [inline subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java:183]: Can we log the exception at `debug` level, just to have the stacktrace kept somewhere?

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java [hunk 3]
```diff
@@ -160,6 +161,32 @@ public File provisionFromArchive(JavaToolchainSpec spec, File jdkArchive, URI ur
...
+        }
+        LOGGER.warn("While provisioning Java toolchain from '{}' to satisfy spec '{}', having metadata '{}', " +
+                "leftover content with metadata '{}' found in install folder '{}'. Existing installation will be replaced by the new download.",
```
Comment (octylFractal) [inline subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java:186]: Suggestion for better flow of the message:
```suggestion
        LOGGER.warn("While provisioning Java toolchain from '{}' to satisfy spec '{}' (with metadata '{}') " +
                "leftover content (with metadata '{}') was found in the install folder '{}'. The existing installation will be replaced by the new download.",
```

**Review by octylFractal — CHANGES_REQUESTED** [review]

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java (hunk 2 above)
Comment (jbartok) [reply subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java:183]: Yes, definitely.

File: subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java (hunk 3 above)
Comment (jbartok) [reply subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/install/JdkCacheDirectory.java:186]: Feedback from native English speakers always welcome.

**Comment by jbartok** [comment]
@ljacomet & @octylFractal , pls. lift your change requests or let me know what else you would like to modify.

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Pull request merged**